File: sdk/ruby/examples/saml-demo/saml_demo.rb
# GGID SAML SSO Demo (Ruby)
# Run: GGID_URL=... ruby saml_demo.rb
require 'webrick'
require_relative '../../lib/ggid/saml'

ggid_url = ENV['GGID_URL'] || 'http://localhost:8080'
entity_id = ENV['SP_ENTITY_ID'] || 'http://localhost:3001/saml/metadata'
acs_url = ENV['ACS_URL'] || 'http://localhost:3001/saml/acs'

server = WEBrick::HTTPServer.new(Port: 3001)

server.mount_proc '/' do |req, res|
  res.body = '<h1>GGID SAML Demo</h1><a href="/login">Login with SAML SSO</a>'
end

server.mount_proc '/saml/metadata' do |req, res|
  res['Content-Type'] = 'application/xml'
  res.body = GGID::SAML.generate_sp_metadata(entity_id: entity_id, acs_url: acs_url)
end

server.mount_proc '/login' do |req, res|
  sso_url = "#{ggid_url}/saml/sso"
  url = GGID::SAML.build_authn_request_url(sso_url: sso_url, entity_id: entity_id, acs_url: acs_url, relay_state: '/profile')
  res.set_redirect(WEBrick::HTTPStatus::Found, url)
end

server.mount_proc '/saml/acs' do |req, res|
  saml_response = req.query['SAMLResponse']
  decoded = Base64.decode64(saml_response)
  res.body = "<h1>SAML ACS</h1><pre>#{decoded}</pre><a href=\"/profile\">Continue</a>"
end

server.mount_proc '/profile' do |req, res|
  res.body = '<h1>Profile</h1><p>Authenticated via SAML SSO</p>'
end

trap('INT') { server.shutdown }
server.start

File: sdk/dart/examples/saml-demo/saml_demo.dart
// GGID SAML SSO Demo (Dart)
// Run: GGID_URL=... dart run saml_demo.dart
import 'dart:io';
import '../../lib/src/saml.dart';

void main() async {
  final ggidUrl = Platform.environment['GGID_URL'] ?? 'http://localhost:8080';
  final entityId = Platform.environment['SP_ENTITY_ID'] ?? 'http://localhost:3001/saml/metadata';
  final acsUrl = Platform.environment['ACS_URL'] ?? 'http://localhost:3001/saml/acs';

  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 3001);
  print('SAML demo on http://localhost:3001');

  await for (final request in server) {
    final path = request.uri.path;
    final response = request.response;

    if (path == '/') {
      response.headers.contentType = ContentType.html;
      response.write('<h1>GGID SAML Demo</h1><a href="/login">Login with SAML SSO</a>');
    } else if (path == '/saml/metadata') {
      response.headers.contentType = ContentType.parse('application/xml');
      response.write(generateSPMetadata(SAMLConfig(entityId: entityId, acsUrl: acsUrl)));
    } else if (path == '/login') {
      final ssoUrl = '$ggidUrl/saml/sso';
      final url = buildAuthnRequestUrl(
        ssoUrl: ssoUrl, entityId: entityId, acsUrl: acsUrl, relayState: '/profile',
      );
      response.statusCode = HttpStatus.movedPermanently;
      response.headers.set(HttpHeaders.locationHeader, url);
    } else if (path == '/saml/acs') {
      response.headers.contentType = ContentType.html;
      response.write('<h1>SAML ACS</h1><p>Received SAML response</p><a href="/profile">Continue</a>');
    } else if (path == '/profile') {
      response.headers.contentType = ContentType.html;
      response.write('<h1>Profile</h1><p>Authenticated via SAML SSO</p>');
    }
    await response.close();
  }
}
